File: app/src/main/java/com/spiderbiggen/randomchampionselector/adapters/ChampionAdapter.java
package com.spiderbiggen.randomchampionselector.adapters;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.ddragon.DDragon;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.model.ImageType;

import java.util.List;

public class ChampionAdapter extends RecyclerView.Adapter<ChampionAdapter.ViewHolder> {

    private static final String TAG = ChampionAdapter.class.getSimpleName();
    private final Context context;

    private final View.OnClickListener clickListener;
    private List<Champion> champions;

    public ChampionAdapter(final Context context, final List<Champion> champions, final View.OnClickListener clickListener) {
        this.context = context;
        this.champions = champions;
        this.clickListener = clickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // create a new view
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_champion_item, parent, false);
        v.setOnClickListener(clickListener);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Champion champion = champions.get(position);

        DDragon dDragon = new DDragon(context);
        Bitmap icon = dDragon.getChampionBitmap(champion, ImageType.SQUARE);
        Bitmap backGround = dDragon.getChampionBitmap(champion, ImageType.SPLASH);
        holder.nameView.setText(champion.getName());
        holder.titleView.setText(champion.getCapitalizedTitle());
        holder.iconView.setImageBitmap(icon);
        holder.backGroundView.setImageBitmap(backGround);
    }

    @Override
    public int getItemCount() {
        return champions != null ? champions.size() : 0;
    }

    public void setChampions(List<Champion> champions) {
        this.champions = champions;
        notifyDataSetChanged();
    }

    public Champion getChampion(int position) {
        if (position > getItemCount() || position < 0) return null;
        return champions.get(position);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        public ImageView backGroundView;
        public ImageView iconView;
        public TextView nameView;
        public TextView titleView;

        public ViewHolder(View itemView) {
            super(itemView);
            nameView = itemView.findViewById(R.id.championName);
            titleView = itemView.findViewById(R.id.championTitle);
            iconView = itemView.findViewById(R.id.championIcon);
            backGroundView = itemView.findViewById(R.id.champion_background);
        }
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/adapters/ChampionAdapter.kt
package com.spiderbiggen.randomchampionselector.adapters

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.spiderbiggen.randomchampionselector.R
import com.spiderbiggen.randomchampionselector.databinding.ListChampionItemBinding
import com.spiderbiggen.randomchampionselector.models.Champion
import com.spiderbiggen.randomchampionselector.util.data.cache.BitmapCache
import com.spiderbiggen.randomchampionselector.util.data.onMainThread
import kotlinx.coroutines.*
import java.io.IOException

/**
 * Defines all behaviour for this [RecyclerView.Adapter].
 *
 * @author Stefan Breetveld
 */
class ChampionAdapter(
    private val clickListener: View.OnClickListener,
    champions: Collection<Champion> = listOf()
) : RecyclerView.Adapter<ChampionAdapter.ViewHolder>() {

    private val champions = champions.toMutableList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        // create a new view
        val v = ListChampionItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        v.root.setOnClickListener(clickListener)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.champion = champions[position]
    }

    override fun getItemCount(): Int = champions.size

    /**
     * Update the list of champions.
     * @param champions a list of [Champion] objects
     */
    fun setChampions(champions: Collection<Champion>) {
        this.champions.clear()
        this.champions.addAll(champions)
        notifyDataSetChanged()
    }

    /**
     * Retrieve the champion stored at that position in the list.
     *
     * @return the requested champion or null if the given index doesn't exist
     */
    fun getChampion(position: Int): Champion? = champions.getOrNull(position)

    /**
     * [ViewHolder] for [ChampionAdapter] handles all changes to the ui when a different champion is set.
     *
     * @author Stefan Breetveld
     */
    class ViewHolder internal constructor(private val binding: ListChampionItemBinding) :
        RecyclerView.ViewHolder(binding.root),
        CoroutineScope by CoroutineScope(Dispatchers.Default) {

        private var a: Job? = null

        internal var champion: Champion? = null
            set(value) {
                if (value == null) throw NullPointerException()
                a?.cancel()
                updateImage(value)
                binding.championName.text = value.name
                binding.championTitle.text = value.capitalizedTitle
                field = value
            }

        private fun updateImage(champion: Champion) {
            a = launch(Dispatchers.IO) {
                val bitmap = try {
                    Log.d("ChampionAdapter", "Image: ${champion.name}")
                    BitmapCache.loadBitmap(champion)
                } catch (e: IOException) {
                    Log.e("ChampionAdapter", "error ${e.message}")
                    null
                }
                onMainThread { binding.championSplash.setImageBitmap(bitmap) }
            }
        }
    }
}
